### PR TITLE
Add DDP tutorial series (video)

### PR DESCRIPTION
**Pull request by suraj813**
Adds the RST pages to go along with the DDP code walkthrough videos

### CODE REVIEW

File: beginner_source/ddp_fault_tolerance.rst [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+So at any failure, you only lose the training progress from the last saved snapshot. 
+
+In elastic training, whenever there are any membership changes (adding or removing nodes), torchrun will terminate and spawn processes
+on available devices. Having this structure ensures your training job can continue without manual intervention.
+
+
+
+
+
+Diff for `multigpu.py <https://github.com/suraj813/distributed-pytorch/blob/main/multigpu.py>`__ v/s `multigpu_torchrun.py <https://github.com/suraj813/distributed-pytorch/blob/main/multigpu_torchrun.py>`__
```
Comment (malfet) [inline beginner_source/ddp_fault_tolerance.rst:101]: Can you please do not reference your personal git repositories, i.e. add `multigpu.py` either to the same PR or submit it to `[pytorch/examples](https://github.com/pytorch/examples)`

File: beginner_source/ddp_fault_tolerance.rst [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+      
+
+
+View the code used in this video: https://github.com/suraj813/distributed-pytorch/blob/main/multigpu_torchrun.py
+
+
+.. raw:: html
+
+   <div style="margin-top:10px; margin-bottom:10px;">
+     <iframe width="560" height="315" src="https://www.youtube.com/embed/9kIvQOiwYzg" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```
Comment (malfet) [inline beginner_source/ddp_fault_tolerance.rst:51]: Just curious, if this is hosted on PyTorch youtube channel or in ones personal file?

File: beginner_source/ddp_multigpu.rst [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+   .. grid-item-card:: :octicon:list-unordered;1em;` Prerequisites
+      :shadow: none
+
+      * High-level overview of `how DDP works  <ddp_theory.html>`__
+      * A machine with multiple GPUs (this tutorial uses an AWS p3.8xlarge instance)
+      * PyTorch `installed <https://pytorch.org/get-started/locally/>`__ with CUDA
+
+
+View the code used in this video: https://github.com/suraj813/distributed-pytorch/blob/main/multigpu.py
```
Comment (malfet) [inline beginner_source/ddp_multigpu.rst:34]: Same as above, why it needs to be in your personal repo?

**Review by malfet — COMMENTED** [review]
Please confirm that videos are (or will be posted) in PyTorches official channel
Also, please add referenced files to the same repo

File: beginner_source/ddp_fault_tolerance.rst (hunk 1 above)
Comment (suraj813) [reply beginner_source/ddp_fault_tolerance.rst:101]: I'll submit to the Examples repo

File: beginner_source/ddp_fault_tolerance.rst (hunk 2 above)
Comment (suraj813) [reply beginner_source/ddp_fault_tolerance.rst:51]: On the PyTorch youtube channel

File: beginner_source/ddp_multigpu.rst (hunk 3 above)
Comment (suraj813) [reply beginner_source/ddp_multigpu.rst:34]: Added the code to the examples repo; PR: https://github.com/pytorch/examples/pull/1065

File: beginner_source/ddp_fault_tolerance.rst [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+Fault-tolerant Distributed Training with ``torchrun``
+=====================================================
+
+Authors: `Suraj Subramanian <https://github.com/suraj813>`__
+
+In distributed training, a single process failure can
+disrupt the entire training job. Since the susceptibility for failure can be higher here, making your training
+script robust is particularly important here. You might also prefer your training job to be *elastic* i.e. the ability to increase or decrease the number
+of processes in realtime while running the training job.
```
Comment (rohan-varma) [inline beginner_source/ddp_fault_tolerance.rst:15]: do we actually provide this realtime capability? i.e. if we set max nodes to 3 and min nodes to 1, and we start training with 1 node, and 1 more node comes alive, do we proactively pause training, re-rendezvous, and continue, or does there need to be some sort of failure that stops the training? 

cc @d4l3k

File: beginner_source/ddp_fault_tolerance.rst [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+PyTorch has a utility ``torchrun`` that provides fault-tolerance as well
+as elastic training. When a failure occurs, torchrun logs the errors and
+attempts to automatically restart all the processes from the last saved
+“snapshot” of the training job. 
+
+The snapshot saves more than just the model state; it can include
+details about the number of epochs run, optimizer states or any other
+mutable attribute of the training job necessary for its continuity.
```
Comment (rohan-varma) [inline beginner_source/ddp_fault_tolerance.rst:24]: s/mutable/stateful maybe

File: beginner_source/ddp_multigpu.rst [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+Constructing the process group
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: diff
+
+   + def ddp_setup(rank, world_size):
+   +   """
+   +   Args:
+   +       rank: Unique identifier of each process
```
Comment (rohan-varma) [inline beginner_source/ddp_multigpu.rst:84]: nit: specify type (int)

File: beginner_source/ddp_multigpu.rst [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+   - ckp = self.model.state_dict()
+   + ckp = self.model.module.state_dict()
+   ...
+   ...
+   - if epoch % self.save_every == 0:
+   + if self.gpu_id == 0 and epoch % self.save_every == 0:
+      self._save_checkpoint(epoch)
+
+We only need to save model checkpoints from one process. Without this
```
Comment (rohan-varma) [inline beginner_source/ddp_multigpu.rst:138]: mention that if this is being done, `_save_checkpoint` should not have any collective calls

File: beginner_source/ddp_multigpu.rst [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+   -  device = 0      # shorthand for cuda:0
+   -  main(device, total_epochs, save_every)
+   +  world_size = torch.cuda.device_count()
+   +  mp.spawn(main, args=(world_size, total_epochs, save_every,), nprocs=world_size)
+
+-  Include new arguments ``rank`` (replacing ``device``) and
+   ``world_size``.
+-  ``rank`` is auto-allocated by DDP when calling
+   ```mp.spawn`` <https://pytorch.org/docs/stable/multiprocessing.html#spawning-subprocesses>`__.
+-  ``world_size`` is the number of processes/GPUs we want to use
```
Comment (rohan-varma) [inline beginner_source/ddp_multigpu.rst:171]: overall we don't encourage / don't support anything other than 1 proc per GPU. so maybe reword as:

world_size is the number of processes across the training job. For GPU training, this corresponds to the number of GPUs in use, and each process works on a dedicated GPU.

File: intermediate_source/ddp_minGPT.rst [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+Saving and Loading from the cloud
+~~~~~~~~~~~~~~~~~~~~~~~~
+In the video, we save training snapshots directly to the cloud. This gives us the flexibility to continue training
+from any node that has access to the cloud bucket.
+
+
+Using Mixed Precision
+~~~~~~~~~~~~~~~~~~~~~~~~
+To speed things up, you might be able to use `Mixed Precision <https://pytorch.org/docs/stable/amp.html>`__ to train your models. 
+In Mixed Precision, some parts of the training process are carried out in FP16 half-precision, while other steps
```
Comment (rohan-varma) [inline intermediate_source/ddp_minGPT.rst:73]: "carried out in reduced precision" as amp now supports other types than just fp17

File: intermediate_source/ddp_minGPT.rst [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+In Mixed Precision, some parts of the training process are carried out in FP16 half-precision, while other steps 
+that are more sensitive to precision drops are maintained in FP32 precision. The `use_amp <https://github.com/suraj813/minGPT-ddp/tree/use_amp>`
+branch contains the code for training with Mixed Precision.
+
+
+When is DDP not enough?
+~~~~~~~~~~~~~~~~~~~~~~~~
+A typical training run's memory footprint consists of model weights, activations, gradients, the input batch, and the optimizer state.
+Since DDP replicates the model on each GPU, it only works when GPUs have sufficient capacity to accomodate the full footprint. 
+When models grow larger, more aggressive techniques like `FSDP <https://pytorch.org/docs/stable/fsdp.html>`__ are required; here the model is not replicated but "sharded" across all the GPUs,
```
Comment (rohan-varma) [inline intermediate_source/ddp_minGPT.rst:82]: maybe also mention activation checkpointing

File: intermediate_source/ddp_minGPT.rst [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+to learn how we trained a 1 Trillion parameter model with FSDP.
+
+
+Further Reading
+---------------
+-  `Multi-node training with DDP <ddp_multinode.html>`__ (previous tutorial in this series)
+-  `Mixed Precision training <https://pytorch.org/docs/stable/amp.html>`__
+-  `Fully-Sharded Data Parallel <https://pytorch.org/docs/stable/fsdp.html>`__
+-  `Training a 1T parameter model with FSDP <https://medium.com/pytorch/training-a-1-trillion-parameter-model-with-pytorch-fully-sharded-data-parallel-on-aws-3ac13aa96cff>`__
+-  Less' FSDP videos
```
Comment (rohan-varma) [inline intermediate_source/ddp_minGPT.rst:93]: Link? Also make this sound more official maybe, such as "FSDP video tutorial series"?

File: intermediate_source/ddp_multinode.rst [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+.. note:: 
+Torchrun supports *heteregenous scaling* i.e. each of your multinode
+machines can have different number of workers participating in the
+training job. In the video, I deployed the code on 2 machines with 4 and
+2 GPUs each.
+
+
+
+Troubleshooting
```
Comment (rohan-varma) [inline intermediate_source/ddp_multinode.rst:68]: In the future, we should add comprehensive documentation onto how to ensure things like RDMA, EFA, etc are being leveraged if applicable

File: intermediate_source/ddp_multinode.rst [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+
+.. raw:: html
+
+   <div style="margin-top:10px; margin-bottom:10px;">
+     <iframe width="560" height="315" src="https://www.youtube.com/embed/KaAJtI1T2x4" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </div>
+
+
+.. note::
```
Comment (rohan-varma) [inline intermediate_source/ddp_multinode.rst:48]: do we need code snippets here?

File: intermediate_source/ddp_multinode.rst [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+   .. grid-item-card:: :octicon:list-unordered;1em;` Prerequisites
+      :shadow: none
+
+      * Familiarity with `multi-GPU training <beginner/ddp_multigpu.html>`__ and `torchrun <beginner/ddp_fault_tolerance.html>`__ 
+      * 2 or more TCP-reachable GPU machines (this tutorial uses AWS p3.2xlarge instances)
+      * PyTorch `installed <https://pytorch.org/get-started/locally/>`__ with CUDA on all machines
+
+
+View the code used in this video: https://github.com/pytorch/examples/tree/main/distributed/ddp-tutorial-series/blob/main/multinode_torchrun.py
```
Comment (rohan-varma) [inline intermediate_source/ddp_multinode.rst:38]: link is broken for me

**Review by rohan-varma — COMMENTED** [review]
Awesome!! Have some initial thoughts / questions

File: intermediate_source/ddp_minGPT.rst (hunk 11 above)
Comment (suraj813) [reply intermediate_source/ddp_minGPT.rst:93]: oops this commit missed the push. thanks for catching!

File: beginner_source/ddp_fault_tolerance.rst (hunk 4 above)
Comment (suraj813) [reply beginner_source/ddp_fault_tolerance.rst:15]: From reading the docs, my sense is that all processes are terminated and restarted on all available nodes with re-rendezvous, to continue from the last available checkpoint. Is "realtime" inaccurate to describe this?

Comment (suraj813) [reply beginner_source/ddp_fault_tolerance.rst:15]: rephrased to "ability to increase or decrease the number of processes during training."

File: beginner_source/ddp_multigpu.rst (hunk 7 above)
Comment (suraj813) [reply beginner_source/ddp_multigpu.rst:138]: Can you explain this more? Why should we not have collective calls here, and what collective calls might typically be needed while saving checkpoints?

File: intermediate_source/ddp_multinode.rst (hunk 14 above)
Comment (suraj813) [reply intermediate_source/ddp_multinode.rst:38]: fixed now

File: beginner_source/ddp_fault_tolerance.rst [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+don't need to. For instance,
+
+-  You don't need to set environment variables or explicitly pass the ``rank`` and ``world_size``; torchrun assigns this along with several other `environment variables <https://pytorch.org/docs/stable/elastic/run.html#environment-variables>`__.
+-  No need to call ``mp.spawn`` in your script; you only need a generic ``main()`` entrypoint, and launch the script with ``torchrun``. This way the same script can be run in non-distributed as well as single-node and multinode setups. 
+-  Gracefully restarting training from the last saved training snapshot
+
+
+Graceful restarts
+~~~~~~~~~~~~~~~~~~~~~
+For graceful restarts, it helps to have the following structure:
```
Comment (svekars) [inline beginner_source/ddp_fault_tolerance.rst:68]:
```suggestion
For graceful restarts, you can use the following structure:
```

File: beginner_source/ddp_multigpu.rst [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+training <../intermediate/ddp_multinode.html>`__ \|\| `mingpt training <../intermediate/ddp_minGPT.html>`__
+
+
+Multi GPU training with DDP
+===========================
+
+Authors: `Suraj Subramanian <https://github.com/suraj813>`__
+
+In the `previous tutorial <ddp_theory.html>`__, we got a high-level overview of how DDP works; now we see how to use DDP in code.
+In this tutorial we start with a single-GPU training script, and migrate that to running it on 4 GPUs on a single node.
```
Comment (svekars) [inline beginner_source/ddp_multigpu.rst:12]:
```suggestion
In this tutorial, we start with a single-GPU training script and migrate that to running it on 4 GPUs on a single node. 
```

File: beginner_source/ddp_multigpu.rst [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+
+Diff for `single_gpu.py <https://github.com/pytorch/examples/blob/main/distributed/ddp-tutorial-series/single_gpu.py>`__ v/s `multigpu.py <https://github.com/pytorch/examples/blob/main/distributed/ddp-tutorial-series/multigpu.py>`__
+----------------------------------------------------
+
+These are the changes you typically make to a single-GPU training script to enable DDP.
+
+Imports
+~~~~~~~
+-  ``torch.multiprocessing`` is a PyTorch wrapper around python's native
```
Comment (svekars) [inline beginner_source/ddp_multigpu.rst:58]:
```suggestion
-  ``torch.multiprocessing`` is a PyTorch wrapper around Python's native
```

File: beginner_source/ddp_multigpu.rst [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+   +   shuffle=False,
+   +   sampler=DistributedSampler(train_dataset),
+   )
+
+
+Saving model checkpoints
+~~~~~~~~~~~~~~~~~~~~~~~~
+-  We only need to save model checkpoints from one process. Without this
+condition, each process would save its copy of the identical mode. Read
+more on `saving and loading models with
```
Comment (svekars) [inline beginner_source/ddp_multigpu.rst:131]:
```suggestion
more on saving and loading models with
```

File: beginner_source/ddp_multigpu.rst [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+DDP <https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#save-and-load-checkpoints>`__
```
Comment (svekars) [inline beginner_source/ddp_multigpu.rst:132]:
```suggestion
DDP `here<https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#save-and-load-checkpoints>`__.
```

File: beginner_source/ddp_series_intro.rst [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+fault-tolerant distributed training.
+
+The tutorial assumes a basic familiarity with model training in PyTorch.
+
+Running the code
+----------------
+
+You will need multiple CUDA GPUs to run the tutorial code. Typically
+this can be done on a cloud instance with multiple GPUs (the tutorials
+use an AWS p3 instance with 4 GPUs).
```
Comment (svekars) [inline beginner_source/ddp_series_intro.rst:33]:
```suggestion
use an Amazon EC2 P3 instance with 4 GPUs).
```

**Review by svekars — COMMENTED** [review]
Looks awesome! I added some editorial suggestions - let me know if you have any questions.

File: beginner_source/ddp_fault_tolerance.rst [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+
+In distributed training, a single process failure can
+disrupt the entire training job. Since the susceptibility for failure can be higher here, making your training
+script robust is particularly important here. You might also prefer your training job to be *elastic* i.e. the ability to increase or decrease the number
+of processes in realtime while running the training job.
+
+PyTorch has a utility ``torchrun`` that provides fault-tolerance as well
+as elastic training. When a failure occurs, torchrun logs the errors and
+attempts to automatically restart all the processes from the last saved
+“snapshot” of the training job.
```
[the remark this answers is not in the document]
Comment (suraj813) [reply beginner_source/ddp_fault_tolerance.rst:20]: snapshot is not a variable or function used in code, so backticks are not appropriate here.

File: intermediate_source/ddp_minGPT.rst [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+that are more sensitive to precision drops are maintained in FP32 precision. 
+
+
+When is DDP not enough?
+~~~~~~~~~~~~~~~~~~~~~~~~
+A typical training run's memory footprint consists of model weights, activations, gradients, the input batch, and the optimizer state.
+Since DDP replicates the model on each GPU, it only works when GPUs have sufficient capacity to accomodate the full footprint. 
+When models grow larger, more aggressive techniques might be useful:
+- `activation checkpointing <https://pytorch.org/docs/stable/checkpoint.html>`__: Instead of saving intermediate activations during the forward pass, the activations are 
+recomputed during the backward pass. In this approach, we do more compute but save on memory footprint.
```
[the remark this answers is not in the document]
Comment (suraj813) [reply intermediate_source/ddp_minGPT.rst:81]: will replace 'do' with 'perform'

File: intermediate_source/ddp_multinode.rst [hunk 23]
```diff
@@ -0,0 +1,80 @@
+`Introduction <../beginner/ddp_series_intro.html>`__ \|\| `What is DDP <../beginner/ddp_theory.html>`__ \|\| `Single-node
+Multi-GPU training <../beginner/ddp_multigpu.html>`__ \|\| `Fault
+Tolerance <../beginner/ddp_fault_tolerance.html>`__ \|\| **Multi-node
+training** \|\| `mingpt training <ddp_minGPT.html>`__
+
+Multinode Training
```
[the remark this answers is not in the document]
Comment (suraj813) [reply intermediate_source/ddp_multinode.rst:6]: Added a note about multinode vs single-node training. In future updates, we can add more details too but at the moment it is close to what the video talks about

File: prototype_source/ios_coreml_workflow.rst [hunk 24]
```diff
@@ -26,7 +26,7 @@ Next, since the Core ML delegate is a prototype feature, let's install the PyTor
 
     pip3 install --pre torch torchvision torchaudio -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
 
-    pip3 install coremltools==5.0b5
+    pip3 install coremltools==5.0b5 protobuf==3.20.1
```
[the remark this answers is not in the document]
Comment (suraj813) [reply prototype_source/ios_coreml_workflow.rst:29]: Yes, arising from an earlier branch merge conflict that was preventing me from pushing to this branch

File: beginner_source/ddp_fault_tolerance.rst [hunk 25]
```diff
@@ -0,0 +1,210 @@
+`Introduction <ddp_series_intro.html>`__ \|\| `What is DDP <ddp_theory.html>`__ \|\| `Single-node
+Multi-GPU training <ddp_multigpu.html>`__ \|\| **Fault
+Tolerance** \|\| `Multi-node
+training <../intermediate/ddp_multinode.html>`__ \|\| `mingpt training <../intermediate/ddp_minGPT.html>`__
```
Comment (svekars) [inline beginner_source/ddp_fault_tolerance.rst:4]: In all files: 
Single-node Multi-GPU training => Single-Node Multi-GPU Training
Multi-node training => Multi-Node Training 
mingpt training => minGPT Training

File: beginner_source/ddp_multigpu.rst (hunk 7 above)
Comment (rohan-varma) [reply beginner_source/ddp_multigpu.rst]: Any code that runs on a specific (or subset) of ranks in a group should not have collectives because collectives would require all ranks to run the collective code.

In _save_checkpoint, sometimes users might have collectives to gather certain states across a rank etc, so that wouldn't work and typically the code would have to be moved to before the `if rank == 0` check.

**Review by rohan-varma — APPROVED** [review]
DDP + multi-node setup details LGTM, thank you!

**Review by malfet — APPROVED** [review]

**Review by malfet — COMMENTED** [review]
Hmm, why `custom.css` changes need to be part of those tutorials? (As they affect other tutorials as well) Can you please factor it out as a separate PR or get a thumbs up from somebody who is responsible for design? cc: @svekars 

Technically changes looks good to me, though it would be good to move from .rst files to .py files

**Comment by svekars** [comment]
@malfet that was actually I who added it there =). We need these changes in this PR.

**Review by svekars — APPROVED** [review]
